p2p/p2p-core/src/client/handshaker/builder.rs

```diff
@@ -0,0 +1,427 @@
+use std::{
+    future::{ready, Ready},
+    marker::PhantomData,
+    task::{Context, Poll},
+};
+
+use futures::{stream, Stream};
+use tower::Service;
+use tracing::Span;
+
+use cuprate_wire::{BasicNodeData, CoreSyncData};
+
+use crate::{
+    client::{handshaker::HandShaker, InternalPeerID},
+    services::{
+        AddressBookRequest, AddressBookResponse, CoreSyncDataRequest, CoreSyncDataResponse,
+        PeerSyncRequest, PeerSyncResponse,
+    },
+    AddressBook, BroadcastMessage, CoreSyncSvc, NetworkZone, PeerSyncSvc, ProtocolRequest,
+    ProtocolResponse,
+};
+
+/// A [`HandShaker`] [`Service`] builder.
+///
+/// This builder applies default values to make usage easier, behaviour and drawbacks of the defaults are documented
+/// on the `with_*` method to change it, for example [`HandshakerBuilder::with_protocol_request_handler`].
+///
+/// If you want to use any network other than mainnet you will need to change the core sync service with
+/// [`HandshakerBuilder::with_core_sync_svc`], see that method for details.
+#[derive(Debug, Clone)]
+pub struct HandshakerBuilder<
```

Comment:
I also added a builder for our handshaker which makes it _a lot_ easier to use `cuprate-p2p-core`